vortex-datafusion/src/lib.rs

```diff
@@ -0,0 +1,323 @@
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        Arc::clone(&self.schema_ref)
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    /// Plan an array scan.
+    ///
+    /// Currently, no pushdown is supported. The array is flattened directly into the nearest
+    /// Arrow-compatible encoding, and we emit a single [RecordBatch] of data.
+    async fn scan(
+        &self,
+        _state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        _limit: Option<usize>,
+    ) -> DFResult<Arc<dyn ExecutionPlan>> {
+        if !filters.is_empty() {
+            return exec_err!("vortex does not support filter pushdown");
+        }
+
+        let partitioning = if let Ok(chunked_array) = ChunkedArray::try_from(&self.array) {
+            Partitioning::RoundRobinBatch(chunked_array.nchunks())
+        } else {
+            Partitioning::UnknownPartitioning(1)
+        };
+
+        let plan_properties = PlanProperties::new(
+            EquivalenceProperties::new(self.schema_ref.clone()),
+            partitioning,
+            ExecutionMode::Bounded,
+        );
+
+        Ok(Arc::new(VortexMemoryExec {
+            array: self.array.clone(),
+            projection: projection.cloned(),
+            plan_properties,
+        }))
+    }
+
+    fn supports_filters_pushdown(
+        &self,
+        filters: &[&Expr],
+    ) -> DFResult<Vec<TableProviderFilterPushDown>> {
+        // TODO(aduffy): add support for filter pushdown
+        Ok(filters
+            .iter()
+            .map(|_| TableProviderFilterPushDown::Unsupported)
+            .collect())
+    }
+}
+
+/// Physical plan node for scans against an in-memory, possibly chunked Vortex Array.
+#[derive(Debug, Clone)]
+struct VortexMemoryExec {
+    array: Array,
+    projection: Option<Vec<usize>>,
+    plan_properties: PlanProperties,
+}
+
+impl DisplayAs for VortexMemoryExec {
+    fn fmt_as(&self, _display_type: DisplayFormatType, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+impl VortexMemoryExec {
+    /// Read a single array chunk from the source as a RecordBatch.
+    fn execute_single_chunk(
+        array: Array,
+        projection: &Option<Vec<usize>>,
+        _context: Arc<TaskContext>,
+    ) -> DFResult<SendableRecordBatchStream> {
+        let data = array
+            .flatten()
+            .map_err(|vortex_error| DataFusionError::Execution(format!("{}", vortex_error)))?
+            .into_array();
+
+        // Construct the RecordBatch by flattening each struct field and transmuting to an ArrayRef.
+        let struct_array = StructArray::try_from(data)
+            .map_err(|_| exec_datafusion_err!("top-level array must be Struct encoding"))?;
+
+        let field_order = if let Some(projection) = projection {
+            projection.clone()
+        } else {
+            (0..struct_array.names().len()).collect()
```

Comment:
Does this work? Or does projection refer to the depth-first column indices? May need to hook into ArrayVisitor to count the total number of columns?

Reply:
From the doc comment on the `scan` method:

```
Projection
If specified, only a subset of columns should be returned, in the order specified. The projection is a set of indexes of the fields in Self::schema.
```

I checked the builtin TableProvider for Parquet files, and it defers to `SchemaRef::project(Vec<usize>)` which basically does the thing I'm doing here.

So I guess datafusion doesn't support skipping nested columns

Reply:
Hmmm... that could be problematic for us!

Reply:
Oh, this is by Influx people

Reply:
Yea, one of the linked issues literally has a statement to the effect of "if we added this it would break IOx"

Reply:
i'm gonna punt this for now, it sounds like if we care about integrating spiral with DF we're going to either need to convince DF to accept PRs or we'll need to change how we represent schemas

Reply:
What we could do is write a function that flattens a struct array such that the column names are complex strings, e.g. `foo.bar.[0].baz`, escaping periods.

Then we could do some fudging in Spiral if/when we need it.